src/async-wrap.cc

```diff
@@ -407,6 +407,50 @@ static void DisablePromiseHook(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+class DestroyParam {
+ public:
+  double asyncId;
+  v8::Persistent<Object> target;
+  v8::Persistent<Object> propBag;
+};
+
+
+void AsyncWrap::WeakCallback(
+  const v8::WeakCallbackInfo<DestroyParam>& info
+) {
+  Environment* env = Environment::GetCurrent(info.GetIsolate());
+  DestroyParam* p = info.GetParameter();
+  Local<Object> propBag = PersistentToLocal(info.GetIsolate(), p->propBag);
+
+  Local<String> key = String::NewFromUtf8(info.GetIsolate(), "destroyed");
+  Local<Value> val = propBag->Get(key);
+  if (!val->BooleanValue()) {
+    PushBackDestroyAsyncId(env, p->asyncId);
+  }
+
+  p->propBag.Reset();
+  delete p;
+}
+
+
+static void RegisterDestroyHook(const FunctionCallbackInfo<Value>& args) {
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsNumber());
+  CHECK(args[2]->IsObject());
+
+
+  Environment* env = Environment::GetCurrent(args);
+
+  DestroyParam* p = new DestroyParam();
+  p->target.Reset(env->isolate(), args[0].As<Object>());
+  p->asyncId = args[1]->NumberValue();
+  p->propBag.Reset(env->isolate(), args[2].As<Object>());
+
+  p->target.SetWeak(
+    p, AsyncWrap::WeakCallback, v8::WeakCallbackType::kParameter);
```

Comment:
Isn't this stuff rather perf-costly?

Reply:
Just checked this with `hrtime` a bit:

It appears that while not fast there's a bigger problem:
The `.As<Object>()` are very slow during instantiation of `AsyncResource`. Just leaving `p->target.Reset(isolate, args[0].As<Object>());` in the Register, moves us from 0.2s for 5MM instantions to 2.4s. Is there any way to avoid the v8 ::Cast for that?

Reply:
`.As<T>` is just an unchecked cast; it essentially compiles down to zero instructions at run-time.  It's the `new` and `.Reset()` calls that are slow.

The `.SetWeak()` call itself isn't very slow but it creates more work for the garbage collector.